### DISCUSSION TITLE
Ejemplo con los argumentos de línea de comandos

### DISCUSSION THREAD

**Discussion by martinvilu** [INGCOM-UNRN-P1/cursada-2023#172] [General de la cátedra]
A continuación un `main` simple que muestra el contenido de `argv`.
```C
#include <stdio.h>

int main(int argc, char* argv[])
{
    int i = 0;
    while (i < argc)
    {
        printf("el argumento %d es <%s>\n", i, argv[i]);
        i = i + 1;
    }
    return 0;
}
```
Y la función `atoi` que se encuentra definida en `stdlib.h` para convertir una cadena de texto a un número entero.

Cuando no es posible hacer la conversión, retornará cero.
```C
#include <stdlib.h>

int atoi(const char *cadena);
```

**Comment by IoachimusRoderici** [comment]
> Cuando no es posible hacer la conversión, retornará cero.

¿Y yo como sé si me devolvió 0 porque el número era 0 o porque no pudo hacer la conversión?

Suena como una mala elección ese código de error.

@martinvilu

**Reply by martinvilu** [reply]
Es una función de la librería estándar 🤷

Para saber si se pudo o no hacer el cambio, es necesario crear una función propia

**Reply by vlzJuan** [reply]
Un poco tarde quizas para la primera entrega, pero una opcion que se me ocurre es hacer, como menciona martin, una funcion propia que compare el string en si y el resultado de atoi. Por ejemplo:

```C
// Toma la linea de codigo tomada desde el archivo como parametro. 
// Retorna 1 si se puede hacer la conversion, o 0 si no puede.

int convertible_a_entero(char* linea[]){

    int ret=0   // Variable de retorno. Empieza en 0, asumiendo que no se pudo
    if(atoi(linea)!=0){    // Si la conversion devuelve algo distinto a 0
        ret=1;                 // Asigno 1 al retorno para decir que se puede hacer la conversion.
    }
    else{                       // Caso contrario, AKA atoi(linea)==0
        if(linea=="0"){        // Si la linea recibida que quiero convertir es, de hecho, el numero 0...
            ret=1;                 // Vuelvo a asignar 1 a la variable de retorno. 
                                      // Con esto, aunque atoi() nos devuelva 0, si lo usamos en conjunto con esta funcion,
                                      // podemos saber que es un 0 'de verdad' y no un fallo de conversion.
        }
    }
    return ret;  // Retorno 0 si no pase por ninguna condicion exitosa, o 1 si la conversion se puede hacer.
}

```

**Reply by IoachimusRoderici** [reply]
> `if(linea=="0")`

Los strings en c no se pueden comparar de esa manera. Igual tampoco serviría porque hay muchos strings distintos que representan el 0 (por ejemplo: `"0"`, `"00"`, `"+0"`, `"-000"`, etc.).

¿tal vez [sscanf](https://en.cppreference.com/w/c/io/fscanf) sea una mejor opción?

**Reply by martinvilu** [reply]
Una función que recorra la cadena, y vaya haciendo las multiplicaciones, convirtiendo cada caracter de dígito.

Lindo ejercicio, por cierto.
```C
int cadena_a_entero_segura(char cadena[], int capacidad, int* numero_interpretado)
```

Hay ejercicios muy interesantes combinados con aritmética de punteros, para el procesamiento de palabras en cadenas.

Porque recuerden que pueden hacer cosas como:
```C
char[] cadena = "abcd1234drc";
char* ptr_cadena = cadena;
numero = cadena_a_entero_segura(ptr_cadena+4, 4);
```
E interpretar 'solo' los números de una cadena más general, siendo que es una cadena segura, debiera de 'quejarse' que no hay terminador, pero no hay razón para que no intente hacer la conversión de todas formas.


*_Edit:_* Un mejor nombre para la función sería `cadena_segura_a_int`

**Comment by martinvilu** [comment]
Comentar en hilos 'contestados' hace que no la vea en la lista de conversaciones pendientes
https://github.com/orgs/INGCOM-UNRN-P1/discussions/172#discussioncomment-7114072

Mejor, usá los tres puntitos:
![imagen](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/56625/48464455-d05c-4132-b114-e9e7dfaaa5a2)

Dale a Copy Link y abrí una pregunta en el espacio de preguntas y respuestas.
